File: backend/assetflow-asset-service/src/main/java/com/assetflow/asset/controller/AssetController.java
package com.assetflow.asset.controller;

import com.assetflow.asset.response.ApiResponse;
import com.assetflow.asset.service.AssetDataService;
import io.jsonwebtoken.Claims;
import java.util.List;
import java.util.Map;
import java.util.Set;
import org.springframework.security.core.Authentication;
import org.springframework.http.HttpStatus;
import org.springframework.web.server.ResponseStatusException;
import org.springframework.web.bind.annotation.*;

@RestController
public class AssetController {
	private final AssetDataService service;

	public AssetController(AssetDataService service) {
		this.service = service;
	}

	private Long company(Authentication authentication) {
		return ((Claims) authentication.getPrincipal()).get("companyId", Long.class);
	}
	private Long user(Authentication authentication) {
		Object subject = ((Claims) authentication.getPrincipal()).getSubject();
		return Long.valueOf(String.valueOf(subject));
	}
	private void assertWriteAccess(String type, Authentication auth) {
		Claims claims = (Claims) auth.getPrincipal();
		String role = claims.get("role", String.class);
		if ("COMPANY_ADMIN".equals(role) || "SUPER_ADMIN".equals(role)) return;
		if ("EMPLOYEE".equals(role) && Set.of("asset-request", "asset-return", "maintenance").contains(type)) return;
		throw new ResponseStatusException(HttpStatus.FORBIDDEN, "You do not have permission to modify this resource");
	}

	@GetMapping("/{type:asset|category|vendor|purchase-order|maintenance|asset-allocation|asset-request|asset-transfer|asset-return|repair-history}")
	public ApiResponse<List<Map<String, Object>>> list(@PathVariable String type, @RequestParam(required = false) String search,
			@RequestParam(defaultValue = "0") int page, @RequestParam(defaultValue = "25") int size, Authentication auth) {
		Claims claims = (Claims) auth.getPrincipal();
		return ApiResponse.ok(service.list(type, company(auth), user(auth), claims.get("role", String.class), search, page, size));
	}

	@GetMapping("/{type:asset|category|vendor|purchase-order|maintenance|asset-allocation|asset-request|asset-transfer|asset-return|repair-history}/{id}")
	public ApiResponse<Map<String, Object>> one(@PathVariable String type, @PathVariable Long id, Authentication auth) {
		Claims claims = (Claims) auth.getPrincipal();
		return ApiResponse.ok(service.one(type, company(auth), user(auth), claims.get("role", String.class), id));
	}

	@PostMapping("/{type:asset|category|vendor|purchase-order|maintenance|asset-allocation|asset-request|asset-transfer|asset-return|repair-history}")
	public ApiResponse<Map<String, Object>> create(@PathVariable String type, @RequestBody Map<String, Object> body,
			Authentication auth) {
		assertWriteAccess(type, auth);
		Claims claims = (Claims) auth.getPrincipal();
		if ("EMPLOYEE".equals(claims.get("role", String.class))) {
			if ("asset-request".equals(type) || "asset-return".equals(type) || "maintenance".equals(type)) {
				body.put("employeeId", user(auth)); body.put("requestedBy", user(auth));
				if ("asset-return".equals(type) || "maintenance".equals(type)) service.ensureEmployeeAsset(company(auth), user(auth), body.get("assetId"));
			}
		}
		if ("asset-allocation".equals(type)) body.putIfAbsent("allocatedBy", user(auth));
		if ("asset-transfer".equals(type) || "asset-return".equals(type)) body.putIfAbsent("requestedBy", user(auth));
		return ApiResponse.ok(service.create(type, company(auth), user(auth), body));
	}

	@PostMapping("/asset/import")
	public ApiResponse<Map<String, Object>> importAssets(@RequestBody List<Map<String, Object>> rows, Authentication auth) {
		assertWriteAccess("asset", auth);
		return ApiResponse.ok(service.importAssets(company(auth), user(auth), rows));
	}

	@PutMapping("/{type:asset|category|vendor|purchase-order|maintenance|asset-allocation|asset-request|asset-transfer|asset-return|repair-history}/{id}")
	public ApiResponse<Map<String, Object>> update(@PathVariable String type, @PathVariable Long id,
			@RequestBody Map<String, Object> body, Authentication auth) {
		assertWriteAccess(type, auth);
		Claims claims = (Claims) auth.getPrincipal();
		if ("EMPLOYEE".equals(claims.get("role", String.class))) { body.put("employeeId", user(auth)); if ("asset-return".equals(type) || "maintenance".equals(type)) service.ensureEmployeeAsset(company(auth), user(auth), body.get("assetId")); }
		if ("asset-request".equals(type) && "APPROVED".equalsIgnoreCase(String.valueOf(body.getOrDefault("status", "")))) body.putIfAbsent("approvedBy", user(auth));
		if ("asset-return".equals(type) && "APPROVED".equalsIgnoreCase(String.valueOf(body.getOrDefault("status", "")))) body.putIfAbsent("approvedBy", user(auth));
		if ("asset-transfer".equals(type) && "APPROVED".equalsIgnoreCase(String.valueOf(body.getOrDefault("status", "")))) body.putIfAbsent("approvedBy", user(auth));
		return ApiResponse.ok(service.update(type, company(auth), user(auth), id, body));
	}

	@DeleteMapping("/{type:asset|category|vendor|purchase-order|maintenance|asset-allocation|asset-request|asset-transfer|asset-return|repair-history}/{id}")
	public ApiResponse<Void> delete(@PathVariable String type, @PathVariable Long id, Authentication auth) {
		assertWriteAccess(type, auth);
		service.delete(type, company(auth), user(auth), id);
		return ApiResponse.message("Deleted successfully");
	}
}

File: dotnet-backend/Assetflow.AssetService/Controllers/AssetController.cs
using System.Security.Claims;
using Assetflow.AssetService.Models;
using Assetflow.AssetService.Services;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Assetflow.AssetService.Controllers
{
    [ApiController]
    [Authorize]
    public class AssetController : ControllerBase
    {
        private readonly AssetDataService _service;
        private static readonly HashSet<string> ValidTypes = new(StringComparer.OrdinalIgnoreCase)
        {
            "asset", "category", "vendor", "purchase-order", "maintenance",
            "asset-allocation", "asset-request", "asset-transfer", "asset-return", "repair-history"
        };

        public AssetController(AssetDataService service)
        {
            _service = service;
        }

        private long CompanyId()
        {
            var claim = User.FindFirst("companyId")?.Value;
            return claim != null ? long.Parse(claim) : 0;
        }

        private long UserId()
        {
            var claim = User.FindFirst(ClaimTypes.NameIdentifier) ?? User.FindFirst("sub");
            return claim != null ? long.Parse(claim.Value) : 0;
        }

        private string Role()
        {
            return User.FindFirst("role")?.Value ?? "";
        }

        private void AssertWriteAccess(string type)
        {
            var role = Role();
            if (role is "COMPANY_ADMIN" or "SUPER_ADMIN") return;
            if (role == "EMPLOYEE" && type is "asset-request" or "asset-return" or "maintenance") return;
            throw new UnauthorizedAccessException("You do not have permission to modify this resource");
        }

        [HttpGet("{type}")]
        public async Task<ActionResult<ApiResponse<IEnumerable<IDictionary<string, object>>>>> List(
            string type,
            [FromQuery] string? search,
            [FromQuery] int page = 0,
            [FromQuery] int size = 25)
        {
            if (!ValidTypes.Contains(type)) return NotFound();
            var data = await _service.ListAsync(type, CompanyId(), UserId(), Role(), search, page, size);
            return Ok(ApiResponse<IEnumerable<IDictionary<string, object>>>.Ok(data));
        }

        [HttpGet("{type}/{id}")]
        public async Task<ActionResult<ApiResponse<IDictionary<string, object>>>> One(string type, long id)
        {
            if (!ValidTypes.Contains(type)) return NotFound();
            var data = await _service.OneAsync(type, CompanyId(), UserId(), Role(), id);
            return Ok(ApiResponse<IDictionary<string, object>>.Ok(data));
        }

        [HttpPost("{type}")]
        public async Task<ActionResult<ApiResponse<IDictionary<string, object>>>> Create(string type, [FromBody] IDictionary<string, object> body)
        {
            if (!ValidTypes.Contains(type)) return NotFound();
            AssertWriteAccess(type);

            if (Role() == "EMPLOYEE")
            {
                if (type is "asset-request" or "asset-return" or "maintenance")
                {
                    body["employeeId"] = UserId();
                    body["requestedBy"] = UserId();
                    if (type is "asset-return" or "maintenance")
                    {
                        body.TryGetValue("assetId", out var assetId);
                        await _service.EnsureEmployeeAssetAsync(CompanyId(), UserId(), assetId);
                    }
                }
            }

            if (type == "asset-allocation")
            {
                if (!body.ContainsKey("allocatedBy")) body["allocatedBy"] = UserId();
            }
            if (type is "asset-transfer" or "asset-return")
            {
                if (!body.ContainsKey("requestedBy")) body["requestedBy"] = UserId();
            }

            var data = await _service.CreateAsync(type, CompanyId(), UserId(), body);
            return Ok(ApiResponse<IDictionary<string, object>>.Ok(data));
        }

        [HttpPost("asset/import")]
        public async Task<ActionResult<ApiResponse<IDictionary<string, object>>>> ImportAssets([FromBody] IEnumerable<IDictionary<string, object>> rows)
        {
            AssertWriteAccess("asset");
            var data = await _service.ImportAssetsAsync(CompanyId(), UserId(), rows);
            return Ok(ApiResponse<IDictionary<string, object>>.Ok(data));
        }

        [HttpPut("{type}/{id}")]
        public async Task<ActionResult<ApiResponse<IDictionary<string, object>>>> Update(string type, long id, [FromBody] IDictionary<string, object> body)
        {
            if (!ValidTypes.Contains(type)) return NotFound();
            AssertWriteAccess(type);

            if (Role() == "EMPLOYEE")
            {
                body["employeeId"] = UserId();
                if (type is "asset-return" or "maintenance")
                {
                    body.TryGetValue("assetId", out var assetId);
                    await _service.EnsureEmployeeAssetAsync(CompanyId(), UserId(), assetId);
                }
            }

            string GetStatus(IDictionary<string, object> dict)
            {
                return dict.TryGetValue("status", out var val) ? val?.ToString() ?? "" : "";
            }

            if (type is "asset-request" or "asset-return" or "asset-transfer")
            {
                if ("APPROVED".Equals(GetStatus(body), StringComparison.OrdinalIgnoreCase))
                {
                    if (!body.ContainsKey("approvedBy")) body["approvedBy"] = UserId();
                }
            }

            var data = await _service.UpdateAsync(type, CompanyId(), UserId(), id, body);
            return Ok(ApiResponse<IDictionary<string, object>>.Ok(data));
        }

        [HttpDelete("{type}/{id}")]
        public async Task<ActionResult<ApiResponse<object>>> Delete(string type, long id)
        {
            if (!ValidTypes.Contains(type)) return NotFound();
            AssertWriteAccess(type);
            await _service.DeleteAsync(type, CompanyId(), UserId(), id);
            return Ok(ApiResponse<object>.WithMessage("Deleted successfully"));
        }
    }
}
